Make torch.conj() a composite function and return self for real tensors

anjali411 opened the pull request:
`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR updates `torch.conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. The documentation states that the returned tensor for a real input shouldn't be mutated. We could perhaps return an immutable tensor for this case in future when that functionality is available (@zdevito @ezyang ). 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43270 Make torch.conj() a composite function and return self for real tensors**

Differential Revision: [D23460493](https://our.internmc.facebook.com/intern/diff/D23460493)

anjali411 committed:
Make torch.conj a composite function

[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj a composite function"




[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj a composite function"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -188,7 +188,15 @@ Tensor imag(const Tensor& self) {
 }
 
 Tensor& conj_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, conj_stub); }
```
What about this one?
What happens if you try to use autograd with it? Does it still raise a nice error?

anjali411 replied:
I don't follow. it's only called through `at::_conj` which has a definition in `derivatives.yaml`. how else can you use it with autograd?

albanD replied:
This one is not called from `_conj` no.
And the user can call it directly by doing `a.conj(out=b)`.
And this should raise an error if either a or b requires grad.

albanD commented on torch/_torch_docs.py:
```diff
@@ -1683,7 +1683,8 @@ def merge_dicts(*dicts):
 
-Computes the element-wise conjugate of the given :attr:`input` tensor.
+Computes the element-wise conjugate of the given :attr:`input` tensor. If :attr:`self` tensor
+has a non-complex dtype, this function returns the :attr:`self` tensor.
```
nit: should be input here not self right?

zou3519 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -188,7 +188,15 @@ Tensor imag(const Tensor& self) {
+
+Tensor _conj(const Tensor& self) { return unary_op_impl(self, at::conj_out); }
+
+Tensor conj(const Tensor& self) {
+  if (!self.is_complex()) {
+    return self;
+  }
+  return at::_conj(self);
+}
```
This seems BC-breaking: previously if a user passed in a real tensor to conj, they'd get back a copy of said tensor. What was the rationale behind this change?

FWIW, in numpy, conj returns a copy of the original tensor if it wasn't complex:
```
>>> x = np.array([0.1, 0.1])
>>> y = np.conj(x)
>>> y.flags
...
OWNDATA : True
```

anjali411 replied:
As discussed in this issue: #41857 switching to the tensorflow style gradients would involve addition of a bunch of `conj` in the gradient definitions. In order to preserve the backward performance for real tensors, it seems like a good option to make `torch.conj` a no op for real tensors.

zou3519 replied:
I see, so the decision in that thread was to go for TF-style gradients. However, TF-style gradients require conj-calls everywhere in the gradient formulas. Because of that, we're trying to optimize calls to conj to avoid regressions.

The downside of calling the operator "fast_conj" is that writers of gradient formulas might use "conj" instead of "fast_conj". I think it is probably OK for us to break BC and to diverge from numpy and have our "conj" not do copies when possible but I am not sure which of these two solutions would be better.

Exploring the design space some more, I noticed that there were some comments about having a conj flag on tensors that reinterprets the tensor as its conjugate (https://github.com/pytorch/pytorch/issues/41857#issuecomment-665129647). Was there a decision on that? That seems like it would eliminate the need to have a `fast_conj`

mruberry replied:
I understand why we want to have a function with this behavior, but have a few questions/suggestions:

- I would prefer our `conj` have the same behavior as NumPy's
- `torch.conj` is, conceptually, a universal function, and all universal functions create new outputs
  - for example, `torch.abs(uint8)` produces a new tensor, even though it could mathematically fulfill its contract by returning self
  - analogously, would we expect `torch.ceil(int64)` to return self or a copy of self? 
- do we expect this function to be used by anyone but developers? Is making a non-user-facing function `self_or_conj` an option to address the scenarios we'd like to target? If we're uncertain about those scenarios, could we start with this?

I appreciate @zou3519's concern about gradient writers mistakenly using `conj` instead of `self_or_conj`, and I agree that's a real possibility, but if we're consistent about using the latter and do our job in code review I think it'll be OK. I'm also less concerned about introducing a fixable performance issue vs. creating an inconsistent and semantically confusing UX for end users.

ezyang replied:
Although I am leaning towards "faithfully replicate numpy behavior" (@anjali411, I think I mentioned that I was having trouble deciding between these two choices in our call), I want to point out some reasons why fast conj by default might make sense.

> do we expect this function to be used by anyone but developers? Is making a non-user-facing function self_or_conj an option to address the scenarios we'd like to target? If we're uncertain about those scenarios, could we start with this?

`fast_conj` absolutely would be used by non-developers. Any time you write code that you want to generalize over both reals and complex you will write conjugations to handle the complex case, and if they are not no-ops, you will unnecessarily slow down the execution of your code in the real case. Some more support from boedekker here https://github.com/pytorch/pytorch/issues/41857#issuecomment-663152126

> Exploring the design space some more, I noticed that there were some comments about having a conj flag on tensors that reinterprets the tensor as its conjugate (#41857 (comment)). Was there a decision on that?

For the BC consideration here, it doesn't help. If we have a fancy version of conjugate that returns a view rather than new tensor, that would still be incompatible with the Numpy behavior. So if you want to keep numpy behavior, you're still going to have to come up with a new name for `view_conj` or whatever.

ezyang replied:
cc @boeddeker

albanD replied:
I would lean towards @ezyang opinion here personally. In particular, "fast_conj absolutely would be used by non-developers." and not having it would mean that people will start adding a lot of `if t.is_complex()` in their code.

@mruberry are there functions for which we explicitly plan to have a different API compared to numpy? And maybe provide a numpy-compliant version in a different namespace? Could this function be one of them?

boeddeker replied:
I am working a lot with numpy and I must say, I never checked if numpy.conj returns a copy or hold the original data.
Sometimes I got the impression, that it is a no op, also for complex numbers, when I measured the execution time (execution time is close to nothing compared to any other function).

Nevertheless, assuming that conj returns a copy in the source tensor, also for real tensors, is a bad idea.

When you introduce `fast_conj`, I would never use `conj`.

In context of torch, I would prefer, when conj never creates a copy.
While the calculation of the copy may be suboptimal, the increased memory consumption is critical.
In numpy you can calculate on demand the conjugate and it will be deleted a few moments later.
In torch, it may be kept for the backward graph. When you call multiple times conj, you store several copies.

So maybe it would also be worth considering, that the conjugate operation gets an argument to enable something like `torch.utils.checkpoint.checkpoint` or have the `conj` flag, that was earlier mentioned.

ezyang replied:
cc @zdevito this may be one of the first cases where we have existentially important laziness on tensors

mruberry replied:
> @mruberry are there functions for which we explicitly plan to have a different API compared to numpy? And maybe provide a numpy-compliant version in a different namespace? 

Yes, but we should strive to limit these discrepancies. In particular we want a compelling reason for the difference (the NumPy behavior would be inconsistent with PyTorch, or it would require an extremely painful deprecation, ...). 

> Could this function be one of them?

Sure. I'm actually less worried about the NumPy compatibility of this function than the PyTorch inconsistency of it. It seems like we're trying to combine the ideas of `torch.conj()` the unary ufunc and a `.conj` attribute with a corresponding view function, like `.real`, `.imag`, and `.T`.

anjali411 committed:
Update on "Make torch.conj a composite function"




[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj a composite function"




[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj a composite function"




[ghstack-poisoned]

gchanan commented:
Here's some fun that I think shows we shouldn't follow NumPy exactly:

```
>>> np.__version__
'1.17.2'

>>> x=np.random.randn(2,3)

>>> x.ctypes.data == x.conj().ctypes.data
True

>>> x.ctypes.data == np.conj(x).ctypes.data
False
```

Perhaps `x.conj()` returning `self` is just a bug, though.  We've had those bugs before.

Same doesn't apply to complex values:

```
y=np.random.randn(2,3) + 1j*np.random.randn(2,3)

>>> y.ctypes.data == y.conj().ctypes.data
False

>>> y.ctypes.data == np.conj(y).ctypes.data
False
```

gchanan commented on test/test_torch.py:
```diff
@@ -19310,6 +19310,12 @@ def _do_transpose(self, x, contiguous=False, dim0=0, dim1=1):
+    def test_conj_self(self, device, dtype):
+        t = torch.ones(5, 5, device=device)
+        s = t.fast_conj()
+        self.assertTrue(s is t)
```
do we actually want this or do we want `s` to be a view of `t`?

anjali411 replied:
we could but I think we should just return same tensor as input with same storage to make it a complete no op

gchanan commented:
I guess my questions are:
1) what is the "right" numpy conj behavior?  I don't expect numpy meant to distinguish the functional and method variant.  @rgommers?
2) is the the semantics of some function we actually want long-term?  I would expect `fast_conj` or `view_conj` or whatever to do something fast/view for complex as well.  I'm unsure if you are planning this, considering this, or you are set on the current semantics.
3) As an API issue, IMO you should start the name with the most "discoverable" information -- think code completion.  People will type `torch.conj...` when looking for conjugate functionality, not start with `fast...`.  So whatever we choose for the name, it should start with `conj`

mruberry commented:
> 1. what is the "right" numpy conj behavior?  I don't expect numpy meant to distinguish the functional and method variant.  @rgommers?

I filed https://github.com/numpy/numpy/issues/17124 about the discrepancy.

rgommers commented:
> what is the "right" numpy conj behavior?  I don't expect numpy meant to distinguish the functional and method variant.  @rgommers?

I commented on the NumPy issue Mike opened. For the decision to be made here, I think it's more important that PyTorch has a consistent design here rather than that it follow NumPy exactly. The copy-vs-view behavior in NumPy can be inconsistent for no good reason, and is very hard to change at this point.

I'd prefer if functions and methods always had the same behaviour.

anjali411 commented:
> Based on this new information (thanks @gchanan for looking, this is really useful information), I've amended my position: I think we should do a serious investigation of doing conjugation view by default; I kind of suspect we still won't do it for `conj` because it is too surprising (because it is just too different from how all of our other unary operations work), but I think we should try to understand it better.

> An alternative to conjugate views is peephole lazy tensors, which @zdevito has advocated for. Peephole lazy tensors work like this:
> 
> * When you write `x.conj()`, we don't immediately do the conjugation operation. Instead, we simply produce a tensor which says, "This is x, but conj." The distinction between this and the conjugate flag, is that this tensor is _semantically_ a new copy of the tensor. So for example, if we then attempted to mutate the result, we would simply directly materialize the conjugated result and then mutate that.
> * If the result is used by another operation which doesn't support fusion, we materialize the conjugated tensor; all subsequent uses of the result make use of this materialized tensor

I agree that we should do a more thorough investigation to add some sort of flag. I like the idea of peephole lazy tensors. `tensor.imag` however would not be a view anymore unless, as you mentioned, we add a negative flag. One option is to only have a negative flag and conjugation would just mean turning on negative flag for the floating point imag tensor (`tensor.imag`).  

Also, I guess writing on a tensor returned from a view (with conj flag turned on) would mean we'll have to first update the entire tensor (even if we were writing on only a small part of the original tensor).

As a temporary resolution, I think we should update `torch.conj` to not return a new tensor with new storage for real tensors. The reasons why I think it's ok to be divergent from numpy:
1. `torch.conj` is not a very useful operation on real tensors (kinda like using `torch.ceil` on long tensors), so I don't see many people using it anyway. We provide enough functionality for people who want numpy like behavior to do that (`if not tensor.is_complex() do ...`). 
2.  @gchanan mentioned jit backward compatibility issue regarding adding a new op and renaming or removing it in future. Simply updating conj also helps there.

There are two choices for what `torch.conj` can return for real tensors:
1. returns self (same tensor and storage as input)
2. view of self (new tensor but same storage as input)

Both seem like reasonable options. I am personally leaning towards `self` to make conj a total no op for real tensors.

anjali411 committed:
Update on "Add a composite function torch.fast_conj()"

`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR adds `torch.fast_conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. This operator is exposed in torch so that people who want this kind of behavior are able to use it.

The name of this function might be a little misleading since it doesn't perform the computations any faster for complex tensors (faster for real tensors but ...). If anyone has name suggestions, please let me know.




[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj() a composite function and return self for real tensors"

`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR updates `torch.conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. The documentation states that the returned tensor for a real input shouldn't be mutated. We could perhaps return an immutable tensor for this case in future when that functionality is available (@zdevito @ezyang ). 




[ghstack-poisoned]

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -1752,7 +1752,8 @@ def merge_dicts(*dicts):
 
-Computes the element-wise conjugate of the given :attr:`input` tensor.
+Computes the element-wise conjugate of the given :attr:`input` tensor. If the input has a non-complex dtype,
+the output tensor shouldn't be mutated.
```
as per @ezyang 's suggestion

ezyang replied:
Actually, since we're exploring conjugate view, let me suggest we make an even stronger claim: we should tell users to flat out NOT mutate the output of this function, ever. If you avoid doing so, you will write code that is forwards compatible with any possible semantics of conj (it always returns a new copy, it returns a new copy sometimes, it always returns a view).

I despair about actually making sure people don't do this with just a doc update. We really need some way to mark views as non-mutable. Maybe this wouldn't even be that hard to do...

anjali411 committed:
Update on "Make torch.conj() a composite function and return self for real tensors"

`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR updates `torch.conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. The documentation states that the returned tensor for a real input shouldn't be mutated. We could perhaps return an immutable tensor for this case in future when that functionality is available (@zdevito @ezyang ). 




[ghstack-poisoned]

anjali411 committed:
Update on "Make torch.conj() a composite function and return self for real tensors"

`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR updates `torch.conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. The documentation states that the returned tensor for a real input shouldn't be mutated. We could perhaps return an immutable tensor for this case in future when that functionality is available (@zdevito @ezyang ). 




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -288,6 +288,10 @@
 
+- func: _conj(Tensor self) -> Tensor
+  use_c10_dispatcher: full
+  variants: function
```
I hope this doesn't get traced; that would be `_convolution` all over again.

albanD replied:
Maybe we want to rename it to `conj_complex()` as it is just a version of `conj()` that **only** accepts complex inputs?

anjali411 replied:
@ezyang sorry I am not aware of the issue caused due to `_convolution`. what was it? 
@albanD yeah that makes sense. will do!

ezyang replied:
Alias information for `conj` above is now not correct; you need to give it an alias annotation similar to what `contiguous` has.

ezyang replied:
@anjali411 Basically, `_convolution`, which has a bunch of extra arguments that people typically don't care about, is what is being traced to serialized models, meaning that now it is very difficult to add new parameters to `_convolution` because they are FC-breaking for all models.

conj is considerably less trafficked so the risks of getting it wrong here are lower, but it still seems to me that we should prefer tracing `conj` and not `_conj`. You should test which one we're getting in the trace.

mruberry replied:
Luckily I think the intent here is to keep _conj's signature equivalent to conj, since it's just an autograd hack to only apply autograd to complex inputs to conj. As you point out, @ezyang, _convolution suffers because it's secretly pulling in arguments the user doesn't expect.

Interesting question for what gets traced when ops are composite though: the composite ops or the original symbol. I don't know offhand.

albanD replied:
Given that the code that handle tracing and autograd was the same until very recently. I expect that only the inner function is traced.

ezyang commented on test/test_torch.py:
```diff
@@ -18669,6 +18669,12 @@ def _do_transpose(self, x, contiguous=False, dim0=0, dim1=1):
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes()))
+    def test_conj_self(self, device, dtype):
+        t = torch.ones(5, 5, device=device)
+        s = t.fast_conj()
```
this needs to be updated

ezyang commented:
In the interest of moving things along, I'd prefer to accept this patch. However, there are a bunch of ways this could go wrong. I'd prefer it if someone else could also ACK.

I'll also take a look at how hard it would be to mark tensors as immutable.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -288,6 +288,10 @@
 
 - func: conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```
Future work: inplace conj_?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1752,7 +1752,8 @@ def merge_dicts(*dicts):
 conj(input, *, out=None) -> Tensor
 
-Computes the element-wise conjugate of the given :attr:`input` tensor.
+Computes the element-wise conjugate of the given :attr:`input` tensor. If the input has a non-complex dtype,
```
Would: "Computes the element-wise complex conjugate of :attr\`input\`." be a clearer fist sentence? "Conjugate" seems overloaded.

I wonder if the second sentence shouldn't be a note and/or a warning? Maybe something like:

NOTE: if :attr\`input\` is a non-complex dtype this function just returns  :attr:`input`.

WARNING: In the future torch.conj may return a non-writeable view of :attr:`input`. It's recommended that programs not modify the tensor returned by torch.conj to be compatible with this change.

anjali411 replied:
I think the behavior for non-complex tensors should go in the description, but added a warning as per your suggestion.

mruberry reviewed (approved):
This approach makes sense. Although this function may change in the future to return an immutable view the complex tensor UX is still in "beta", and @ezyang's suggestion to clarify the documentation seems like a reasonable mitigation. Few cleanup requirements:

- test needs update (per @ezyang's note)
- doc needs clarification

As for NumPy Compatibility, sometimes we do need to be different than NumPy for reasons of performance, our focus on neural networks, or our support of autograd, and I think we have a consensus that this approach is the best way (for now) to implement performant complex autograd.

From the perspective of how PyTorch's functions work, it's important that users know what to expect when they call a function. That is, ideally they should be able to infer what the function will do without reading its documentation. This means that functions should be named reasonably (e.g. torch.linalg.outer computes an outer product, and torch.ger is indecipherable for most users), and the system should be so consistent that users don't need to memorize a lot of exceptions or strange rules (for example, functions and methods in PyTorch perform the same computations). 

torch.conj is a weird duck because there are good arguments for it being a unary ufunc like it is in NumPy, a view function, or a function like contiguous, cpu, cuda, and to. The first class of functions has always created a new tensor. The second usually returns a view (*ahem*, reshape), and the third class, which is hard to name, has returned self or a new tensor. From our discussion, however, it seems like torch.conj will eventually be in one of the latter two groups. Maybe the third class should be extended  (and maybe torch.abs(uint8) should return self) or maybe torch.conj will eventually return an (immutable) view. Either way, this current approach seems like our best approximation of that future behavior.

ezyang commented:
Thanks Mike!

anjali411 committed:
Update on "Make torch.conj() a composite function and return self for real tensors"

`torch.conj` is a very commonly used operator for complex tensors, but it's mathematically a no op for real tensors. Switching to tensorflow gradients for complex tensors (as discussed in #41857) would involve adding `torch.conj()` to the backward definitions for a lot of operators. In order to preserve autograd performance for real tensors and maintain numpy compatibility for `torch.conj`, this PR updates `torch.conj()` which behaves the same for complex tensors but performs a view/returns `self` tensor for tensors of non-complex dtypes. The documentation states that the returned tensor for a real input shouldn't be mutated. We could perhaps return an immutable tensor for this case in future when that functionality is available (@zdevito @ezyang ). 




[ghstack-poisoned]
